### PR TITLE
Fixes a sampling bug affecting QubitDevice

### PR DESCRIPTION
**Pull request by josh146**
**Description of the Change:**

* Fixes a bug in the new `QubitDevice` class, where the function `generate_samples()` was generating the incorrect samples for a given device probability.

* Fixes a bug in the new `QubitDevice` class, where the marginal probability was not taking into account the ordering of the wires to be marginalized.

* Removes the wire sorting and permutations happening in:
  
  - `Tensor.eigvals`
  - `Hermitian.eigendecomposition`
  - `QubitDevice.sample`

  These permutations were workarounds that are no longer necessary now that the bug in marginal probs has been found.

* Adds additional tests

* Modifies the tensor expectation/variance tests, to ensure they pass in both analytic and non-analytic mode.

**Benefits:** n/a

**Possible Drawbacks:** n/a

**Related GitHub Issues:** n/a

### CODE REVIEW

File: tests/test_qubit_device.py [hunk 1]
```diff
@@ -363,53 +367,47 @@ def test_correct_conversion_two_states(self, mock_qubit_device, monkeypatch):
                             [
                             (
                               np.array([2, 3, 2, 0, 0]),
-                              np.array([[0, 1, 0, 0],
-                                        [1, 1, 0, 0],
-                                        [0, 1, 0, 0],
-                                        [0, 0, 0, 0],
-                                        [0, 0, 0, 0]],
+                              np.array([[1, 0],
```
Comment (antalszava) [inline tests/test_qubit_device.py]: The comments at 363-365 will also be false.

File: tests/ops/test_qubit_ops.py [hunk 2]
```diff
@@ -151,31 +151,6 @@ def test_hermitian_eigegendecomposition_multiple_wires(
...
-
-        key = tuple(observable.flatten().tolist())
-        assert np.allclose(qml.Hermitian._eigs[key]["eigval"], eigvals, atol=tol, rtol=0)
-        assert np.allclose(qml.Hermitian._eigs[key]["eigvec"], eigvecs, atol=tol, rtol=0)
-        assert len(qml.Hermitian._eigs) == 1
-
-        # Reverse the order of the wires to check if a new entry was added
-        eigendecomp = qml.Hermitian(observable, wires=list(range(num_wires))[::-1]).eigendecomposition_of_permuted
-
-        assert len(qml.Hermitian._eigs) == 2
```
Comment (josh146) [inline tests/ops/test_qubit_ops.py]: This test is no longer required, as the permutation no longer occurs inside `Hermitian.eigvals`. As a result, calling Hermitian with different wire orderings will no longer create a new key in the matrix!

File: tests/test_qubit_device.py (hunk 1 above)
Comment (josh146) [reply tests/test_qubit_device.py]: Removed!

File: tests/test_qubit_device.py [hunk 3]
```diff
@@ -309,15 +309,15 @@ def test_auxiliary_methods_called_correctly(self, mock_qubit_device, monkeypatch
         """Tests that the generate_samples method calls on its auxiliary methods correctly"""
 
         mock_qubit_device._wires_measured = [1,2]
```
Comment (antalszava) [inline tests/test_qubit_device.py]:
```suggestion
```

File: pennylane/_qubit_device.py [hunk 4]
```diff
@@ -19,6 +19,7 @@
 # e.g. instead of expval(self, observable, wires, par) have expval(self, observable)
```
Comment (antalszava) [inline pennylane/_qubit_device.py]: The ``_wires_measured`` attribute can be removed as is from ``_qubit_device``: Lines 75, 88, 118 would not be relevant in ``execute()``

Comment (antalszava) [reply pennylane/_qubit_device.py]: This would affect ``test_qubit_device.py`` lines 288, 298

File: tests/test_hermitian_edge_cases.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        re_b = A[0, 1].real
+        d = A[1, 1]
+
+        expected = ((a - d) * np.cos(theta) + 2 * re_b * np.sin(theta) * np.sin(phi) + a + d) / 2
+        assert np.allclose(res, expected, atol=0.01, rtol=0)
+
+    @pytest.mark.parametrize("w1, w2", list(itertools.permutations(range(4), 2)))
+    def test_hermitian_two_wires_permuted(self, w1, w2, analytic, theta, phi, varphi):
+        """Test that an hermitian expectation with various wires permuted works"""
+        dev = qml.device("default.qubit", wires=4, shots=1000000, analytic=False)
```
Comment (antalszava) [inline tests/test_hermitian_edge_cases.py]:
```suggestion
        dev = qml.device("default.qubit", wires=4, shots=1000000, analytic=analytic)
```

File: tests/test_hermitian_edge_cases.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        res = circuit()
+
+        a = A[0, 0]
+        re_b = A[0, 1].real
+        d = A[1, 1]
+
+        expected = ((a - d) * np.cos(theta) + 2 * re_b * np.sin(theta) * np.sin(phi) + a + d) / 2
+        assert np.allclose(res, expected, atol=0.01, rtol=0)
+
+    @pytest.mark.parametrize("w1, w2", list(itertools.permutations(range(4), 2)))
```
Comment (antalszava) [inline tests/test_hermitian_edge_cases.py]: Great test! :wink:

File: pennylane/_qubit_device.py [hunk 7]
```diff
@@ -306,11 +307,17 @@ def marginal_prob(self, prob, wires=None):
...
+        if wires is None:
+            # no need to marginalize
+            return prob
+
         wires = np.hstack(wires)
         inactive_wires = list(set(range(self.num_wires)) - set(wires))
         prob = prob.reshape([2] * self.num_wires)
-        return np.apply_over_axes(np.sum, prob, inactive_wires).flatten()
+
+        basis_states = np.array(list(itertools.product([0, 1], repeat=len(wires))))
```
Comment (antalszava) [inline pennylane/_qubit_device.py]: Adding something like ``# Computing permutation indices based on the wires specified``  might be nice here

Comment (antalszava) [reply pennylane/_qubit_device.py]: And adding a similar disclaimer in the docstring as well

File: pennylane/_qubit_device.py (hunk 4 above)
Comment (josh146) [reply pennylane/_qubit_device.py]: done!

File: pennylane/_qubit_device.py (hunk 7 above)
Comment (josh146) [reply pennylane/_qubit_device.py]: done!

File: tests/test_hermitian_edge_cases.py (hunk 5 above)
Comment (josh146) [reply tests/test_hermitian_edge_cases.py]: oh no, good catch :tired_face:

**Pull request merged**